one_off_scripts/backfill_daily_test_rollups.py

```diff
@@ -0,0 +1,191 @@
+    flaky_fail_count: int
+
+    sum_duration_seconds: float
+    last_duration_seconds: float
+
+    latest_run: datetime
+
+    commits_where_fail: set[str] = field(default_factory=set)
+
+
+def get_test_analytics_repos(start_repoid):
+    # get all repos that have test_analytics_enabled == True
+    test_analytics_repos = Repository.objects.filter(
+        test_analytics_enabled=True
+    ).order_by("repoid")
+
+    if start_repoid is not None:
+        test_analytics_repos = test_analytics_repos.filter(repoid__gt=start_repoid)
+
+    return test_analytics_repos
+
+
+def process_instance(
+    rollup_dict: dict[tuple[str, str], RollupObj],
+    flake_dict: dict[str, list[tuple[datetime, datetime | None]]],
+    instance: TestInstance,
+):
+    pass_count = 0
+    fail_count = 0
+    skip_count = 0
+    flaky_fail_count = 0
+    duration_seconds = 0
+    created_at = instance.created_at
+    commitid = instance.commitid
+
+    match instance.outcome:
+        case Outcome.Pass:
+            pass_count = 1
+        case Outcome.Skip:
+            skip_count = 1
+        case _:
+            fail_count = 1
+            if (flaky_range_list := flake_dict.get(instance.test_id)) is not None:
+                for range in flaky_range_list:
+                    if range[0] <= instance.created_at and (
+                        range[1] is None or instance.created_at < range[1]
+                    ):
+                        flaky_fail_count += 1
+                        break
+
+    if (entry := rollup_dict.get((instance.test_id, instance.branch))) is not None:
+        entry.pass_count += pass_count
+        entry.fail_count += fail_count
+        entry.skip_count += skip_count
+        entry.flaky_fail_count += flaky_fail_count
+        entry.sum_duration_seconds += duration_seconds
+        entry.last_duration_seconds = duration_seconds
+        entry.latest_run = created_at
+        entry.commits_where_fail.add(commitid)
+
+    else:
+        rollup_dict[(instance.test_id, instance.branch)] = RollupObj(
+            pass_count,
+            fail_count,
+            skip_count,
+            flaky_fail_count,
+            duration_seconds,
+            duration_seconds,
+            created_at,
+            {commitid},
+        )
+
+
+def save_rollups(rollup_dict, repoid, date):
+    for obj_key, obj in rollup_dict.items():
+        rollup = DailyTestRollup(
+            repoid=repoid,
+            date=date,
+            test_id=obj_key[0],
+            branch=obj_key[1],
+            pass_count=obj.pass_count,
+            fail_count=obj.fail_count,
+            skip_count=obj.skip_count,
+            flaky_fail_count=obj.flaky_fail_count,
+            commits_where_fail=list(obj.commits_where_fail),
+            latest_run=obj.latest_run,
+            last_duration_seconds=obj.last_duration_seconds,
+            avg_duration_seconds=obj.sum_duration_seconds
+            / (obj.pass_count + obj.fail_count),
+        )
+
+        rollup.save()
```

Comment:
django has a [`bulk_create`](https://docs.djangoproject.com/en/5.1/ref/models/querysets/#bulk-create) method as well, though I recommend to always define a reasonable chunk size.